File: bin/6002_m_vpc_mut_ty_log2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
封装 m_vpc_mut_ty_log2 因子计算：
- 从 config.ini 读取路径与窗口参数
- 仅负责滑动拼接固定窗口数据
- 调用 m_vpc_mut_ty_log2 原算子输出结果
"""

import configparser
import os
import sys
import time
from datetime import datetime

import numpy as np
import pandas as pd

from m_vpc_mut_ty_log2_factor import m_vpc_mut_ty_log2


class SimpleDatabase:
    """封装原算子所需的数据接口"""

    def __init__(self, amt: pd.DataFrame, close: pd.DataFrame):
        self.depend_data = {
            "FactorData.Basic_factor.amt_minute": amt,
            "FactorData.Basic_factor.close_adj_minute": close,
        }


def read_config(config_path: str) -> dict:
    """读取配置文件，仅提取必要字段"""
    parser = configparser.ConfigParser()
    parser.read(config_path, encoding="utf-8")

    section = "6002_m_vpc_mut_ty_log2"
    if not parser.has_section(section):
        raise ValueError(f"配置文件缺少节: {section}")

    result = {}
    str_fields = {
        "input_amt": "input_amt",
        "input_close": "input_close",
        "output_csv": "output_csv",
    }
    int_fields = {
        "precision": "precision",
        "lag": "lagWindow",
        "step_size": "stepSize",
    }

    for key, option in str_fields.items():
        if not parser.has_option(section, option):
            raise ValueError(f"配置文件缺少必需项: [{section}] {option}")
        result[key] = parser.get(section, option)

    for key, option in int_fields.items():
        if not parser.has_option(section, option):
            raise ValueError(f"配置文件缺少必需项: [{section}] {option}")
        result[key] = parser.getint(section, option)

    return result


def read_csv_data(file_path: str) -> pd.DataFrame:
    """读取无表头CSV数据"""
    print(f"正在读取: {file_path}")
    data = pd.read_csv(file_path, header=None, dtype=float)
    print(f"  维度: {data.shape[0]} x {data.shape[1]}")
    return data


def format_value(value: float, precision: int) -> str:
    """输出格式化数值（兼容 NaN）"""
    if pd.isna(value) or np.isnan(value):
        return "nan"
    return f"{value:.{precision}f}"


def main():
    try:
        script_dir = os.path.dirname(os.path.abspath(__file__))
        config_path = os.path.join(script_dir, "config.ini")

        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        print(f"[{timestamp}] === 6002_m_vpc_mut_ty_log2 因子计算（Python封装） ===")
        print(f"[{timestamp}] 正在读取配置文件: {config_path}")

        config = read_config(config_path)

        # 读取输入数据
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] 正在读取输入数据...")
        amt = read_csv_data(config["input_amt"])
        close = read_csv_data(config["input_close"])

        if amt.shape != close.shape:
            raise ValueError(f"输入数据维度不一致: amt {amt.shape}, close {close.shape}")

        num_timepoints, num_stocks = amt.shape
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] 数据维度: {num_timepoints} x {num_stocks}")

        lag_window = config["lag"]
        step_size = config["step_size"]
        precision = config["precision"]

        if num_timepoints < lag_window:
            raise ValueError(f"时间点数({num_timepoints})小于窗口大小({lag_window})")

        # 因子实例仅负责计算，不再做配置解析（原始代码不接受参数）
        factor = m_vpc_mut_ty_log2()

        # 输出路径追加 _py
        original_output_path = config["output_csv"]
        if "." in os.path.basename(original_output_path):
            name, ext = os.path.splitext(original_output_path)
            output_path = name + "_py" + ext
        else:
            output_path = original_output_path + "_py"

        output_dir = os.path.dirname(output_path)
        if output_dir and not os.path.exists(output_dir):
            os.makedirs(output_dir, exist_ok=True)

        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] 配置参数:")
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}]   lagWindow = {lag_window}")
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}]   stepSize = {step_size}")
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}]   precision = {precision}")

        with open(output_path, "w", encoding="utf-8") as output_file:
            output_file.write(f"# Generated at: {datetime.now().strftime('%Y-%m-%d %H:%M')}\n")
            header = "time_index,step,operation,elapsed_ms"
            for stock_idx in range(num_stocks):
                header += f",stock_{stock_idx}_mut_ty_log2"
            output_file.write(header + "\n")

            # 初始化阶段：取前 lagWindow 个时间点
            print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] 开始初始化...")
            init_start = time.time()
            init_amt = amt.iloc[0:lag_window, :]
            init_close = close.iloc[0:lag_window, :]
            init_db = SimpleDatabase(init_amt, init_close)
            init_result = factor.calc_single(init_db)
            init_elapsed = (time.time() - init_start) * 1000
            print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] 初始化完成，耗时: {init_elapsed:.3f} ms")

            init_row = f"{lag_window - 1},0,init,{init_elapsed:.3f}"
            for stock_idx in range(num_stocks):
                value = init_result.iloc[stock_idx] if stock_idx < len(init_result) else np.nan
                init_row += f",{format_value(value, precision)}"
            output_file.write(init_row + "\n")

            # 更新阶段：按 stepSize 滑动窗口
            print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] 开始持续更新...")
            update_count = 0
            current_time = lag_window

            while current_time < num_timepoints:
                update_size = min(step_size, num_timepoints - current_time)
                window_end = current_time + update_size
                window_start = max(0, window_end - lag_window)

                window_amt = amt.iloc[window_start:window_end, :]
                window_close = close.iloc[window_start:window_end, :]
                database = SimpleDatabase(window_amt, window_close)

                update_start = time.time()
                update_result = factor.calc_single(database)
                update_elapsed = (time.time() - update_start) * 1000
                update_count += 1

                time_index = window_end - 1
                update_row = f"{time_index},{update_count},update,{update_elapsed:.3f}"
                for stock_idx in range(num_stocks):
                    value = update_result.iloc[stock_idx] if stock_idx < len(update_result) else np.nan
                    update_row += f",{format_value(value, precision)}"
                output_file.write(update_row + "\n")

                print(
                    f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] "
                    f"更新 #{update_count}: 时间点 {time_index}/{num_timepoints-1}, "
                    f"耗时: {update_elapsed:.3f} ms"
                )

                current_time += update_size

        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] === 计算完成 ===")
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] 结果已保存到: {output_path}")

    except Exception as exc:
        print(f"错误: {exc}", file=sys.stderr)
        raise


if __name__ == "__main__":
    main()

File: src/factor_case/6002_m_vpc_mut_ty_log2.cpp
#include <iostream>
#include <iomanip>
#include <fstream>
#include <chrono>
#include <vector>
#include <algorithm>
#include <climits>
#include <cmath>

// 项目头文件
#include "../Tool/config_reader.h"
#include "../Tool/timestamp.h"
#include "../Tool/database.h"
#include "../Tool/profiler.h"
#include "../Factor/MVpcMutTyLog/m_vpc_mut_ty_log2.h"
#include "../Eigen_extra/Eigen_extra.h"

using Eigen::MatrixXd;
using Eigen::VectorXd;
using std::cout;
using std::endl;
using std::vector;

// 辅助函数：计算耗时（毫秒）
double getElapsedMs(const std::chrono::high_resolution_clock::time_point& start) {
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    return duration.count() / 1000.0;
}

int main() {
    try {
        // 读取配置
        Tool::ConfigReader config("config.ini");

        std::string input_amt = config.getString("6002_m_vpc_mut_ty_log2", "input_amt", "");
        std::string input_close = config.getString("6002_m_vpc_mut_ty_log2", "input_close", "");
        std::string output_csv = config.getString("6002_m_vpc_mut_ty_log2", "output_csv", "");

        int precision = config.getInt("6002_m_vpc_mut_ty_log2", "precision", 6);
        int stepSize = config.getInt("6002_m_vpc_mut_ty_log2", "stepSize", 1);
        int lagWindow = config.getInt("6002_m_vpc_mut_ty_log2", "lagWindow", 120);

        if (input_amt.empty() || input_close.empty() || output_csv.empty()) {
            std::cerr << "错误: 配置文件中缺少必要的输入/输出路径" << std::endl;
            return 1;
        }

        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] === 6002_m_vpc_mut_ty_log2 因子计算 ===" << endl;
        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] 配置参数:" << endl;
        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "]   lagWindow = " << lagWindow << endl;
        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "]   stepSize = " << stepSize << endl;
        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "]   precision = " << precision << endl;

        // 加载输入数据
        Tool::MemoryDatabase database;
        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] 正在加载输入数据..." << endl;
        {
            PROFILE_SCOPE("数据读取");
            database.loadFromCSVParallel({
                {"FactorData.Basic_factor.amt_minute", input_amt},
                {"FactorData.Basic_factor.close_adj_minute", input_close}
            });
        }

        MatrixXd amt = database.getMatrix("FactorData.Basic_factor.amt_minute");
        MatrixXd close = database.getMatrix("FactorData.Basic_factor.close_adj_minute");

        int num_timepoints = static_cast<int>(amt.rows());
        int num_stocks = static_cast<int>(amt.cols());

        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] 数据维度: " << num_timepoints << " x " << num_stocks << endl;

        if (close.rows() != num_timepoints || close.cols() != num_stocks) {
            std::cerr << "错误: 收盘价数据维度不匹配: " << close.rows() << " x " << close.cols() << endl;
            return 1;
        }

        if (num_timepoints < lagWindow) {
            std::cerr << "错误: 时间点数(" << num_timepoints << ")小于初始化窗口大小(" << lagWindow << ")" << std::endl;
            return 1;
        }

        m_vpc_mut_ty_log2 factor;

        std::ofstream output_file(output_csv);
        if (!output_file.is_open()) {
            std::cerr << "错误: 无法创建输出文件: " << output_csv << std::endl;
            return 1;
        }

        output_file << "# Generated at: " << Tool::Timestamp::getCurrentTimestamp() << "\n";
        output_file << "time_index,step,operation,elapsed_ms";
        for (int i = 0; i < num_stocks; ++i) {
            output_file << ",stock_" << i << "_mut_ty_log2";
        }
        output_file << "\n";

        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] 开始初始化..." << endl;
        auto init_start = std::chrono::high_resolution_clock::now();

        Ma initAmt = amt.block(0, 0, lagWindow, num_stocks);
        Ma initClose = close.block(0, 0, lagWindow, num_stocks);

        {
            PROFILE_SCOPE("m_vpc_mut_ty_log2::Init");
            int ret = factor.Init(initAmt, initClose);
            if (ret != 0) {
                std::cerr << "错误: Init 返回非零状态码: " << ret << std::endl;
                return 1;
            }
        }

        double init_elapsed = getElapsedMs(init_start);
        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] 初始化完成，耗时: " 
             << std::fixed << std::setprecision(3) << init_elapsed << " ms" << endl;

        const Ve& init_value = factor.getValue();
        output_file << lagWindow - 1 << ",0,init," << std::fixed << std::setprecision(3) << init_elapsed;
        for (int i = 0; i < num_stocks; ++i) {
            if (i < init_value.size()) {
                double value = init_value[i];
                if (std::isnan(value)) {
                    output_file << ",nan";
                } else {
                    output_file << "," << std::fixed << std::setprecision(precision) << value;
                }
            } else {
                output_file << ",nan";
            }
        }
        output_file << "\n";

        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] 开始持续更新..." << endl;
        int update_count = 0;
        int current_time = lagWindow;

        while (current_time < num_timepoints) {
            int update_size = std::min(stepSize, num_timepoints - current_time);

            Ma updateAmt = amt.block(current_time, 0, update_size, num_stocks);
            Ma updateClose = close.block(current_time, 0, update_size, num_stocks);

            auto update_start = std::chrono::high_resolution_clock::now();

            {
                PROFILE_SCOPE("m_vpc_mut_ty_log2::Update");
                factor.Update(updateAmt, updateClose);
            }

            double update_elapsed = getElapsedMs(update_start);
            update_count++;

            const Ve& update_value = factor.getValue();
            output_file << (current_time + update_size - 1) << "," << update_count << ",update,"
                        << std::fixed << std::setprecision(3) << update_elapsed;
            for (int i = 0; i < num_stocks; ++i) {
                if (i < update_value.size()) {
                    double value = update_value[i];
                    if (std::isnan(value)) {
                        output_file << ",nan";
                    } else {
                        output_file << "," << std::fixed << std::setprecision(precision) << value;
                    }
                } else {
                    output_file << ",nan";
                }
            }
            output_file << "\n";

            if (update_count % 100 == 0) {
                cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] 更新进度: "
                     << update_count << " 次, 当前时间点: "
                     << (current_time + update_size - 1) << "/" << num_timepoints
                     << ", 本次耗时: "
                     << std::fixed << std::setprecision(3) << update_elapsed << " ms" << endl;
            }

            current_time += update_size;
        }

        factor.Finish();
        output_file.close();

        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] === 计算完成 ===" << endl;
        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] 总更新次数: " << update_count << endl;
        cout << "[" << Tool::Timestamp::getCurrentTimestamp() << "] 结果已保存到: " << output_csv << endl;

        Tool::Profiler::getInstance().printReport();

        std::string perf_report_csv = output_csv;
        size_t last_dot = perf_report_csv.find_last_of(".");
        if (last_dot != std::string::npos) {
            perf_report_csv = perf_report_csv.substr(0, last_dot) + "_performance.csv";
        } else {
            perf_report_csv += "_performance.csv";
        }
        Tool::Profiler::getInstance().exportToCSV(perf_report_csv);

    } catch (const std::exception& e) {
        std::cerr << "错误: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}


